pubsub/tests/unit/subscriber_test.py

```diff
@@ -0,0 +1,469 @@
+        cache.ack_deadline = 55.0
+        await cache.refresh()
+        assert cache.ack_deadline == 55.0
+        assert cache.last_refresh
+
+    @pytest.mark.asyncio
+    async def test_ack_deadline_cache_get_calls_refresh_first_time(
+        subscriber_client
+    ):
+        cache = AckDeadlineCache(
+            subscriber_client, 'fake_subscription', 1)
+        assert await cache.get() == 42
+        assert cache.last_refresh
+
+    @pytest.mark.asyncio
+    async def test_ack_deadline_cache_get_no_call_if_not_outdated(
+        subscriber_client
+    ):
+        cache = AckDeadlineCache(
+            subscriber_client, 'fake_subscription', 1000)
+        cache.ack_deadline = 33
+        cache.last_refresh = time.perf_counter()
+        assert await cache.get() == 33
+        subscriber_client.get_subscription.assert_not_called()
+
+    @pytest.mark.asyncio
+    async def test_ack_deadline_cache_get_no_call_first_time_if_not_outdated(
+        subscriber_client
+    ):
+        cache = AckDeadlineCache(
+            subscriber_client, 'fake_subscription', 1000)
+        cache.last_refresh = time.perf_counter()
+        assert await cache.get() == float('inf')
+        subscriber_client.get_subscription.assert_not_called()
+
+    @pytest.mark.asyncio
+    async def test_ack_deadline_cache_get_refreshes_if_outdated(
+            subscriber_client):
+        cache = AckDeadlineCache(
+            subscriber_client, 'fake_subscription', 0)
+        cache.ack_deadline = 33
+        assert await cache.get() == 42
+        assert cache.last_refresh
+        subscriber_client.get_subscription.assert_called_once()
+
+    @pytest.mark.asyncio
+    async def test_ack_deadline_cache_first_get_failed(subscriber_client):
+        f = asyncio.Future()
+        f.set_exception(RuntimeError)
+        subscriber_client.get_subscription = MagicMock(
+            return_value=f)
+        cache = AckDeadlineCache(
+            subscriber_client, 'fake_subscription', 10)
+        assert await cache.get() == float('inf')
+        assert cache.last_refresh
+        subscriber_client.get_subscription.assert_called_once()
+
+    # ========
+    # producer
+    # ========
+
+    @pytest.mark.asyncio
+    async def test_producer_fetches_messages(subscriber_client):
+        queue = asyncio.Queue()
+        producer_task = asyncio.ensure_future(
+            producer(
+                'fake_subscription',
+                queue,
+                subscriber_client,
+                max_messages=1,
+                metrics_client=MagicMock()
+            )
+        )
+        message, pulled_at = await asyncio.wait_for(queue.get(), 0.1)
+        producer_task.cancel()
+        assert message
+        assert isinstance(pulled_at, float)
+
+    @pytest.mark.asyncio
+    async def test_producer_timeout_error_is_ok(subscriber_client):
+        async def f(*_args, **_kwargs):
+            await asyncio.sleep(0)
+            raise RuntimeError
```

Comment:
I'm a bit confused with this test. It sounds like we are trying to test that if `subscriber_client.pull` raises `TimeoutError` then we don't raise the exception. But in the test we mock it to raise `RuntimeError` which would have a different codepath than the `TimeoutError` case.

In fact, I was curious and just ran the test locally, and it _seems_ that somehow this raises `CancelledError`. Could you confirm this test is doing what it is supposed to be doing? 😄

Reply:
You absolutely right. There was also mock object missing so we didn't really know if anything was called at all. Please check now!